File: web-client/src/services/tablePartMemoryOptimizer.ts
/**
 * Table Part Memory Optimizer
 * Manages memory usage for large table datasets
 */

interface MemoryOptimizationOptions {
  maxMemoryMB: number
  cleanupThreshold: number
  compressionEnabled: boolean
  cacheStrategy: 'lru' | 'lfu' | 'fifo'
  monitoringInterval: number
}

interface MemoryStats {
  usedMemoryMB: number
  maxMemoryMB: number
  cacheSize: number
  compressionRatio: number
  cleanupCount: number
  lastCleanup: Date
}

interface CacheEntry<T> {
  key: string
  data: T
  compressed?: Uint8Array
  size: number
  accessCount: number
  lastAccessed: Date
  created: Date
}

export class TablePartMemoryOptimizer<T = any> {
  private cache = new Map<string, CacheEntry<T>>()
  private options: MemoryOptimizationOptions
  private stats: MemoryStats
  private monitoringTimer?: number
  private compressionWorker?: Worker

  constructor(options: Partial<MemoryOptimizationOptions> = {}) {
    this.options = {
      maxMemoryMB: 100,
      cleanupThreshold: 0.8,
      compressionEnabled: true,
      cacheStrategy: 'lru',
      monitoringInterval: 5000,
      ...options
    }

    this.stats = {
      usedMemoryMB: 0,
      maxMemoryMB: this.options.maxMemoryMB,
      cacheSize: 0,
      compressionRatio: 0,
      cleanupCount: 0,
      lastCleanup: new Date()
    }

    this.startMonitoring()
    this.initializeCompression()
  }

  /**
   * Store data in optimized cache
   */
  async store(key: string, data: T): Promise<void> {
    const size = this.estimateSize(data)
    
    // Check if we need cleanup before storing
    if (this.shouldCleanup(size)) {
      await this.cleanup()
    }

    const entry: CacheEntry<T> = {
      key,
      data,
      size,
      accessCount: 1,
      lastAccessed: new Date(),
      created: new Date()
    }

    // Compress if enabled and data is large enough
    if (this.options.compressionEnabled && size > 10240) { // 10KB threshold
      try {
        entry.compressed = await this.compress(data)
        entry.size = entry.compressed.length
        // Clear original data to save memory
        entry.data = null as any
      } catch (error) {
        console.warn('Compression failed, storing uncompressed:', error)
      }
    }

    this.cache.set(key, entry)
    this.updateStats()
  }

  /**
   * Retrieve data from cache
   */
  async get(key: string): Promise<T | null> {
    const entry = this.cache.get(key)
    if (!entry) {
      return null
    }

    // Update access statistics
    entry.accessCount++
    entry.lastAccessed = new Date()

    // Decompress if needed
    if (entry.compressed && !entry.data) {
      try {
        entry.data = await this.decompress(entry.compressed)
      } catch (error) {
        console.error('Decompression failed:', error)
        this.cache.delete(key)
        return null
      }
    }

    return entry.data
  }

  /**
   * Check if key exists in cache
   */
  has(key: string): boolean {
    return this.cache.has(key)
  }

  /**
   * Remove entry from cache
   */
  delete(key: string): boolean {
    const deleted = this.cache.delete(key)
    if (deleted) {
      this.updateStats()
    }
    return deleted
  }

  /**
   * Clear all cache
   */
  clear(): void {
    this.cache.clear()
    this.updateStats()
  }

  /**
   * Get current memory statistics
   */
  getStats(): MemoryStats {
    return { ...this.stats }
  }

  /**
   * Force cleanup of cache
   */
  async cleanup(): Promise<void> {
    const targetSize = this.options.maxMemoryMB * this.options.cleanupThreshold
    const currentSize = this.stats.usedMemoryMB

    if (currentSize <= targetSize) {
      return
    }

    const entries = Array.from(this.cache.entries())
    let removedSize = 0
    let removedCount = 0

    // Sort entries based on cache strategy
    const sortedEntries = this.sortEntriesForCleanup(entries)

    // Remove entries until we're under the target size
    for (const [key, entry] of sortedEntries) {
      if (currentSize - removedSize <= targetSize) {
        break
      }

      removedSize += entry.size / (1024 * 1024) // Convert to MB
      removedCount++
      this.cache.delete(key)
    }

    this.stats.cleanupCount++
    this.stats.lastCleanup = new Date()
    this.updateStats()

    console.log(`Memory cleanup: removed ${removedCount} entries, freed ${removedSize.toFixed(2)}MB`)
  }

  /**
   * Optimize memory usage by compressing uncompressed entries
   */
  async optimizeMemory(): Promise<void> {
    if (!this.options.compressionEnabled) {
      return
    }

    const entries = Array.from(this.cache.values())
    let optimizedCount = 0
    let savedMemory = 0

    for (const entry of entries) {
      if (!entry.compressed && entry.data && entry.size > 10240) {
        try {
          const originalSize = entry.size
          entry.compressed = await this.compress(entry.data)
          entry.size = entry.compressed.length
          entry.data = null as any

          savedMemory += (originalSize - entry.size) / (1024 * 1024)
          optimizedCount++
        } catch (error) {
          console.warn('Failed to compress entry:', error)
        }
      }
    }

    this.updateStats()

    if (optimizedCount > 0) {
      console.log(`Memory optimization: compressed ${optimizedCount} entries, saved ${savedMemory.toFixed(2)}MB`)
    }
  }

  /**
   * Preload data with memory-aware batching
   */
  async preloadData(
    keys: string[],
    loadFn: (key: string) => Promise<T>,
    batchSize: number = 10
  ): Promise<void> {
    const batches = this.createBatches(keys, batchSize)

    for (const batch of batches) {
      // Check memory before each batch
      if (this.shouldCleanup()) {
        await this.cleanup()
      }

      // Load batch in parallel
      const loadPromises = batch
        .filter(key => !this.has(key))
        .map(async key => {
          try {
            const data = await loadFn(key)
            await this.store(key, data)
          } catch (error) {
            console.error(`Failed to preload ${key}:`, error)
          }
        })

      await Promise.all(loadPromises)

      // Small delay between batches to prevent blocking
      await new Promise(resolve => setTimeout(resolve, 10))
    }
  }

  /**
   * Destroy optimizer and cleanup resources
   */
  destroy(): void {
    if (this.monitoringTimer) {
      clearInterval(this.monitoringTimer)
    }

    if (this.compressionWorker) {
      this.compressionWorker.terminate()
    }

    this.clear()
  }

  // Private methods

  private startMonitoring(): void {
    this.monitoringTimer = window.setInterval(() => {
      this.updateStats()
      
      // Auto-cleanup if threshold exceeded
      if (this.shouldCleanup()) {
        this.cleanup().catch(error => {
          console.error('Auto-cleanup failed:', error)
        })
      }
    }, this.options.monitoringInterval)
  }

  private initializeCompression(): void {
    if (this.options.compressionEnabled && typeof Worker !== 'undefined') {
      try {
        // Create compression worker for background processing
        const workerCode = `
          self.onmessage = function(e) {
            const { action, data, id } = e.data
            
            try {
              if (action === 'compress') {
                const compressed = new TextEncoder().encode(JSON.stringify(data))
                self.postMessage({ id, result: compressed })
              } else if (action === 'decompress') {
                const decompressed = JSON.parse(new TextDecoder().decode(data))
                self.postMessage({ id, result: decompressed })
              }
            } catch (error) {
              self.postMessage({ id, error: error.message })
            }
          }
        `
        
        const blob = new Blob([workerCode], { type: 'application/javascript' })
        this.compressionWorker = new Worker(URL.createObjectURL(blob))
      } catch (error) {
        console.warn('Failed to create compression worker:', error)
      }
    }
  }

  private async compress(data: T): Promise<Uint8Array> {
    if (this.compressionWorker) {
      return new Promise((resolve, reject) => {
        const id = Math.random().toString(36)
        
        const handler = (e: MessageEvent) => {
          if (e.data.id === id) {
            this.compressionWorker!.removeEventListener('message', handler)
            if (e.data.error) {
              reject(new Error(e.data.error))
            } else {
              resolve(e.data.result)
            }
          }
        }
        
        this.compressionWorker.addEventListener('message', handler)
        this.compressionWorker.postMessage({ action: 'compress', data, id })
      })
    } else {
      // Fallback to synchronous compression
      return new TextEncoder().encode(JSON.stringify(data))
    }
  }

  private async decompress(compressed: Uint8Array): Promise<T> {
    if (this.compressionWorker) {
      return new Promise((resolve, reject) => {
        const id = Math.random().toString(36)
        
        const handler = (e: MessageEvent) => {
          if (e.data.id === id) {
            this.compressionWorker!.removeEventListener('message', handler)
            if (e.data.error) {
              reject(new Error(e.data.error))
            } else {
              resolve(e.data.result)
            }
          }
        }
        
        this.compressionWorker.addEventListener('message', handler)
        this.compressionWorker.postMessage({ action: 'decompress', data: compressed, id })
      })
    } else {
      // Fallback to synchronous decompression
      return JSON.parse(new TextDecoder().decode(compressed))
    }
  }

  private estimateSize(data: T): number {
    // Rough estimation of object size in bytes
    const jsonString = JSON.stringify(data)
    return new Blob([jsonString]).size
  }

  private shouldCleanup(additionalSize: number = 0): boolean {
    const projectedSize = this.stats.usedMemoryMB + (additionalSize / (1024 * 1024))
    return projectedSize > (this.options.maxMemoryMB * this.options.cleanupThreshold)
  }

  private sortEntriesForCleanup(entries: [string, CacheEntry<T>][]): [string, CacheEntry<T>][] {
    switch (this.options.cacheStrategy) {
      case 'lru':
        return entries.sort(([, a], [, b]) => 
          a.lastAccessed.getTime() - b.lastAccessed.getTime()
        )
      
      case 'lfu':
        return entries.sort(([, a], [, b]) => 
          a.accessCount - b.accessCount
        )
      
      case 'fifo':
        return entries.sort(([, a], [, b]) => 
          a.created.getTime() - b.created.getTime()
        )
      
      default:
        return entries
    }
  }

  private updateStats(): void {
    let totalSize = 0
    let compressedSize = 0
    let uncompressedSize = 0

    for (const entry of this.cache.values()) {
      totalSize += entry.size
      
      if (entry.compressed) {
        compressedSize += entry.size
        // Estimate original size for compression ratio
        uncompressedSize += entry.size * 2 // Rough estimate
      } else {
        uncompressedSize += entry.size
      }
    }

    this.stats.usedMemoryMB = totalSize / (1024 * 1024)
    this.stats.cacheSize = this.cache.size
    this.stats.compressionRatio = uncompressedSize > 0 ? compressedSize / uncompressedSize : 0
  }

  private createBatches<U>(items: U[], batchSize: number): U[][] {
    const batches: U[][] = []
    for (let i = 0; i < items.length; i += batchSize) {
      batches.push(items.slice(i, i + batchSize))
    }
    return batches
  }
}

/**
 * Factory function to create a memory optimizer
 */
export function createMemoryOptimizer<T>(
  options?: Partial<MemoryOptimizationOptions>
): TablePartMemoryOptimizer<T> {
  return new TablePartMemoryOptimizer<T>(options)
}

File: src/services/table_part_memory_optimizer.py
"""
Table Part Memory Optimizer

Manages memory usage for large table datasets in PyQt6 applications.
Provides caching, compression, and cleanup strategies to handle large datasets efficiently.
"""

import json
import gzip
import time
import threading
from typing import Dict, List, Any, Optional, Tuple, Union
from dataclasses import dataclass, field
from enum import Enum
from PyQt6.QtCore import QObject, pyqtSignal, QTimer
import logging

logger = logging.getLogger(__name__)


class CacheStrategy(Enum):
    """Cache eviction strategies"""
    LRU = "lru"  # Least Recently Used
    LFU = "lfu"  # Least Frequently Used
    FIFO = "fifo"  # First In, First Out


@dataclass
class MemoryOptimizationOptions:
    """Configuration for memory optimization"""
    max_memory_mb: float = 100.0
    cleanup_threshold: float = 0.8
    compression_enabled: bool = True
    cache_strategy: CacheStrategy = CacheStrategy.LRU
    monitoring_interval_ms: int = 5000
    compression_threshold_bytes: int = 10240  # 10KB


@dataclass
class MemoryStats:
    """Memory usage statistics"""
    used_memory_mb: float = 0.0
    max_memory_mb: float = 0.0
    cache_size: int = 0
    compression_ratio: float = 0.0
    cleanup_count: int = 0
    last_cleanup: Optional[float] = None


@dataclass
class CacheEntry:
    """Cache entry with metadata"""
    key: str
    data: Any = None
    compressed_data: Optional[bytes] = None
    size_bytes: int = 0
    access_count: int = 0
    last_accessed: float = field(default_factory=time.time)
    created: float = field(default_factory=time.time)
    is_compressed: bool = False


class TablePartMemoryOptimizer(QObject):
    """
    Memory optimizer for table part data.
    
    Features:
    - Intelligent caching with configurable eviction strategies
    - Data compression for large entries
    - Automatic cleanup when memory thresholds are exceeded
    - Performance monitoring and statistics
    """
    
    # Signals
    memoryStatsUpdated = pyqtSignal(object)  # MemoryStats
    cleanupPerformed = pyqtSignal(int, float)  # removed_count, freed_mb
    compressionCompleted = pyqtSignal(str, float)  # key, compression_ratio
    memoryThresholdExceeded = pyqtSignal(float)  # current_usage_mb
    
    def __init__(self, options: Optional[MemoryOptimizationOptions] = None):
        super().__init__()
        
        self.options = options or MemoryOptimizationOptions()
        self.cache: Dict[str, CacheEntry] = {}
        self.stats = MemoryStats(max_memory_mb=self.options.max_memory_mb)
        
        # Monitoring timer
        self.monitoring_timer = QTimer()
        self.monitoring_timer.timeout.connect(self._update_stats)
        self.monitoring_timer.start(self.options.monitoring_interval_ms)
        
        # Compression lock for thread safety
        self._compression_lock = threading.Lock()
        
        logger.info(f"Memory optimizer initialized with {self.options.max_memory_mb}MB limit")
    
    def store(self, key: str, data: Any) -> bool:
        """
        Store data in optimized cache.
        
        Args:
            key: Unique identifier for the data
            data: Data to store
            
        Returns:
            True if stored successfully, False otherwise
        """
        try:
            # Estimate size
            size_bytes = self._estimate_size(data)
            
            # Check if cleanup is needed
            if self._should_cleanup(size_bytes):
                self.cleanup()
            
            # Create cache entry
            entry = CacheEntry(
                key=key,
                data=data,
                size_bytes=size_bytes,
                access_count=1
            )
            
            # Compress if enabled and data is large enough
            if (self.options.compression_enabled and 
                size_bytes > self.options.compression_threshold_bytes):
                
                compressed_data = self._compress_data(data)
                if compressed_data and len(compressed_data) < size_bytes:
                    entry.compressed_data = compressed_data
                    entry.size_bytes = len(compressed_data)
                    entry.is_compressed = True
                    entry.data = None  # Clear original data to save memory
                    
                    compression_ratio = len(compressed_data) / size_bytes
                    self.compressionCompleted.emit(key, compression_ratio)
                    
                    logger.debug(f"Compressed {key}: {size_bytes} -> {len(compressed_data)} bytes "
                               f"({compression_ratio:.2%})")
            
            # Store in cache
            self.cache[key] = entry
            self._update_stats()
            
            logger.debug(f"Stored {key} ({entry.size_bytes} bytes, compressed: {entry.is_compressed})")
            return True
            
        except Exception as e:
            logger.error(f"Failed to store {key}: {e}")
            return False
    
    def get(self, key: str) -> Optional[Any]:
        """
        Retrieve data from cache.
        
        Args:
            key: Unique identifier for the data
            
        Returns:
            Cached data or None if not found
        """
        entry = self.cache.get(key)
        if not entry:
            return None
        
        try:
            # Update access statistics
            entry.access_count += 1
            entry.last_accessed = time.time()
            
            # Decompress if needed
            if entry.is_compressed and entry.compressed_data:
                if entry.data is None:
                    entry.data = self._decompress_data(entry.compressed_data)
                return entry.data
            
            return entry.data
            
        except Exception as e:
            logger.error(f"Failed to retrieve {key}: {e}")
            # Remove corrupted entry
            self.delete(key)
            return None
    
    def has(self, key: str) -> bool:
        """Check if key exists in cache"""
        return key in self.cache
    
    def delete(self, key: str) -> bool:
        """Remove entry from cache"""
        if key in self.cache:
            del self.cache[key]
            self._update_stats()
            logger.debug(f"Deleted {key} from cache")
            return True
        return False
    
    def clear(self):
        """Clear all cache entries"""
        self.cache.clear()
        self._update_stats()
        logger.info("Cache cleared")
    
    def cleanup(self) -> Tuple[int, float]:
        """
        Perform cache cleanup to free memory.
        
        Returns:
            Tuple of (removed_count, freed_mb)
        """
        target_size_mb = self.options.max_memory_mb * self.options.cleanup_threshold
        current_size_mb = self.stats.used_memory_mb
        
        if current_size_mb <= target_size_mb:
            return 0, 0.0
        
        # Get entries sorted by eviction strategy
        entries = list(self.cache.items())
        sorted_entries = self._sort_entries_for_cleanup(entries)
        
        removed_count = 0
        freed_bytes = 0
        
        # Remove entries until we're under the target size
        for key, entry in sorted_entries:
            if current_size_mb - (freed_bytes / (1024 * 1024)) <= target_size_mb:
                break
            
            freed_bytes += entry.size_bytes
            removed_count += 1
            del self.cache[key]
        
        freed_mb = freed_bytes / (1024 * 1024)
        
        # Update statistics
        self.stats.cleanup_count += 1
        self.stats.last_cleanup = time.time()
        self._update_stats()
        
        # Emit signals
        self.cleanupPerformed.emit(removed_count, freed_mb)
        
        logger.info(f"Cleanup completed: removed {removed_count} entries, freed {freed_mb:.2f}MB")
        return removed_count, freed_mb
    
    def optimize_memory(self) -> Tuple[int, float]:
        """
        Optimize memory usage by compressing uncompressed entries.
        
        Returns:
            Tuple of (optimized_count, saved_mb)
        """
        if not self.options.compression_enabled:
            return 0, 0.0
        
        optimized_count = 0
        saved_bytes = 0
        
        for entry in list(self.cache.values()):
            if (not entry.is_compressed and 
                entry.data is not None and 
                entry.size_bytes > self.options.compression_threshold_bytes):
                
                try:
                    compressed_data = self._compress_data(entry.data)
                    if compressed_data and len(compressed_data) < entry.size_bytes:
                        original_size = entry.size_bytes
                        
                        entry.compressed_data = compressed_data
                        entry.size_bytes = len(compressed_data)
                        entry.is_compressed = True
                        entry.data = None  # Clear original data
                        
                        saved_bytes += (original_size - entry.size_bytes)
                        optimized_count += 1
                        
                        compression_ratio = len(compressed_data) / original_size
                        self.compressionCompleted.emit(entry.key, compression_ratio)
                        
                except Exception as e:
                    logger.warning(f"Failed to compress {entry.key}: {e}")
        
        saved_mb = saved_bytes / (1024 * 1024)
        
        if optimized_count > 0:
            self._update_stats()
            logger.info(f"Memory optimization: compressed {optimized_count} entries, "
                       f"saved {saved_mb:.2f}MB")
        
        return optimized_count, saved_mb
    
    def preload_data(
        self, 
        keys: List[str], 
        load_fn: callable, 
        batch_size: int = 10
    ):
        """
        Preload data with memory-aware batching.
        
        Args:
            keys: List of keys to preload
            load_fn: Function to load data for a key
            batch_size: Number of items to load per batch
        """
        batches = [keys[i:i + batch_size] for i in range(0, len(keys), batch_size)]
        
        for batch in batches:
            # Check memory before each batch
            if self._should_cleanup():
                self.cleanup()
            
            # Load batch
            for key in batch:
                if not self.has(key):
                    try:
                        data = load_fn(key)
                        self.store(key, data)
                    except Exception as e:
                        logger.error(f"Failed to preload {key}: {e}")
            
            # Small delay between batches
            time.sleep(0.01)
    
    def get_stats(self) -> MemoryStats:
        """Get current memory statistics"""
        return self.stats
    
    def get_cache_info(self) -> Dict[str, Any]:
        """Get detailed cache information"""
        total_size = sum(entry.size_bytes for entry in self.cache.values())
        compressed_count = sum(1 for entry in self.cache.values() if entry.is_compressed)
        
        return {
            'total_entries': len(self.cache),
            'compressed_entries': compressed_count,
            'total_size_mb': total_size / (1024 * 1024),
            'compression_ratio': compressed_count / len(self.cache) if self.cache else 0,
            'cache_strategy': self.options.cache_strategy.value,
            'max_memory_mb': self.options.max_memory_mb
        }
    
    def cleanup_resources(self):
        """Cleanup resources and stop monitoring"""
        self.monitoring_timer.stop()
        self.clear()
        logger.info("Memory optimizer resources cleaned up")
    
    # Private methods
    
    def _estimate_size(self, data: Any) -> int:
        """Estimate size of data in bytes"""
        try:
            # Convert to JSON string and estimate size
            json_str = json.dumps(data, default=str)
            return len(json_str.encode('utf-8'))
        except Exception:
            # Fallback estimation
            return len(str(data).encode('utf-8'))
    
    def _compress_data(self, data: Any) -> Optional[bytes]:
        """Compress data using gzip"""
        try:
            with self._compression_lock:
                json_str = json.dumps(data, default=str)
                return gzip.compress(json_str.encode('utf-8'))
        except Exception as e:
            logger.warning(f"Compression failed: {e}")
            return None
    
    def _decompress_data(self, compressed_data: bytes) -> Any:
        """Decompress data using gzip"""
        with self._compression_lock:
            json_str = gzip.decompress(compressed_data).decode('utf-8')
            return json.loads(json_str)
    
    def _should_cleanup(self, additional_size_bytes: int = 0) -> bool:
        """Check if cleanup is needed"""
        projected_size_mb = (
            self.stats.used_memory_mb + 
            (additional_size_bytes / (1024 * 1024))
        )
        threshold_mb = self.options.max_memory_mb * self.options.cleanup_threshold
        return projected_size_mb > threshold_mb
    
    def _sort_entries_for_cleanup(
        self, 
        entries: List[Tuple[str, CacheEntry]]
    ) -> List[Tuple[str, CacheEntry]]:
        """Sort entries based on cache strategy"""
        if self.options.cache_strategy == CacheStrategy.LRU:
            return sorted(entries, key=lambda x: x[1].last_accessed)
        elif self.options.cache_strategy == CacheStrategy.LFU:
            return sorted(entries, key=lambda x: x[1].access_count)
        elif self.options.cache_strategy == CacheStrategy.FIFO:
            return sorted(entries, key=lambda x: x[1].created)
        else:
            return entries
    
    def _update_stats(self):
        """Update memory statistics"""
        total_size_bytes = sum(entry.size_bytes for entry in self.cache.values())
        compressed_size_bytes = sum(
            entry.size_bytes for entry in self.cache.values() 
            if entry.is_compressed
        )
        
        self.stats.used_memory_mb = total_size_bytes / (1024 * 1024)
        self.stats.cache_size = len(self.cache)
        
        if total_size_bytes > 0:
            self.stats.compression_ratio = compressed_size_bytes / total_size_bytes
        else:
            self.stats.compression_ratio = 0.0
        
        # Check for threshold exceeded
        if self.stats.used_memory_mb > self.options.max_memory_mb:
            self.memoryThresholdExceeded.emit(self.stats.used_memory_mb)
        
        # Emit updated stats
        self.memoryStatsUpdated.emit(self.stats)


def create_memory_optimizer(
    options: Optional[MemoryOptimizationOptions] = None
) -> TablePartMemoryOptimizer:
    """Factory function to create a memory optimizer"""
    return TablePartMemoryOptimizer(options)